File: SudoLinux/models/simple_action.py
"""Preset system shortcuts that can be assigned to macro pad buttons in simple mode."""

from enum import Enum


class SimpleAction(Enum):
    """Linux system shortcuts mapped to macro pad buttons."""

    TAKE_SCREENSHOT = "take_screenshot"
    TAKE_SCREENSHOT_AREA = "take_screenshot_area"
    COPY = "copy"
    PASTE = "paste"
    UNDO = "undo"
    REDO = "redo"
    SAVE = "save"
    SELECT_ALL = "select_all"
    NEW_TAB = "new_tab"
    CLOSE_TAB = "close_tab"
    SWITCH_APP = "switch_app"
    SEARCH = "search"
    WORKSPACES = "workspaces"
    SHOW_DESKTOP = "show_desktop"
    LOCK_SCREEN = "lock_screen"

    @property
    def display_name(self):
        """Human-readable name for display."""
        return {
            SimpleAction.TAKE_SCREENSHOT:      "Screenshot",
            SimpleAction.TAKE_SCREENSHOT_AREA: "Screenshot Area",
            SimpleAction.COPY:                 "Copy",
            SimpleAction.PASTE:                "Paste",
            SimpleAction.UNDO:                 "Undo",
            SimpleAction.REDO:                 "Redo",
            SimpleAction.SAVE:                 "Save",
            SimpleAction.SELECT_ALL:           "Select All",
            SimpleAction.NEW_TAB:              "New Tab",
            SimpleAction.CLOSE_TAB:            "Close Tab",
            SimpleAction.SWITCH_APP:           "Switch App",
            SimpleAction.SEARCH:               "Search",
            SimpleAction.WORKSPACES:           "Workspaces",
            SimpleAction.SHOW_DESKTOP:         "Show Desktop",
            SimpleAction.LOCK_SCREEN:          "Lock Screen",
        }[self]

    @property
    def category(self):
        """Category grouping for display."""
        system = {
            SimpleAction.TAKE_SCREENSHOT, SimpleAction.TAKE_SCREENSHOT_AREA,
            SimpleAction.SEARCH, SimpleAction.WORKSPACES,
            SimpleAction.SHOW_DESKTOP, SimpleAction.LOCK_SCREEN,
            SimpleAction.SWITCH_APP,
        }
        editing = {
            SimpleAction.COPY, SimpleAction.PASTE,
            SimpleAction.UNDO, SimpleAction.REDO,
            SimpleAction.SAVE, SimpleAction.SELECT_ALL,
        }
        if self in system:
            return "System"
        if self in editing:
            return "Editing"
        return "Navigation"

    @property
    def key_combo(self):
        """xdotool key string for this shortcut.

        Returns a string suitable for `xdotool key <combo>`.
        """
        return {
            SimpleAction.TAKE_SCREENSHOT:      "Print",
            SimpleAction.TAKE_SCREENSHOT_AREA: "shift+Print",
            SimpleAction.COPY:                 "ctrl+c",
            SimpleAction.PASTE:                "ctrl+v",
            SimpleAction.UNDO:                 "ctrl+z",
            SimpleAction.REDO:                 "ctrl+shift+z",
            SimpleAction.SAVE:                 "ctrl+s",
            SimpleAction.SELECT_ALL:           "ctrl+a",
            SimpleAction.NEW_TAB:              "ctrl+t",
            SimpleAction.CLOSE_TAB:            "ctrl+w",
            SimpleAction.SWITCH_APP:           "alt+Tab",
            SimpleAction.SEARCH:               "super",
            SimpleAction.WORKSPACES:           "super+Tab",
            SimpleAction.SHOW_DESKTOP:         "super+d",
            SimpleAction.LOCK_SCREEN:          "super+l",
        }[self]

    @staticmethod
    def categories():
        """All categories in display order."""
        return ["System", "Editing", "Navigation"]

    @classmethod
    def actions_in_category(cls, category):
        """Return actions filtered by category."""
        return [a for a in cls if a.category == category]

    @classmethod
    def grouped_by_category(cls):
        """Return actions grouped by category as list of (category, [actions])."""
        return [(cat, cls.actions_in_category(cat)) for cat in cls.categories()]

File: Sudo/Sources/Sudo/Models/SimpleAction.swift
import Foundation
import CoreGraphics

/// Preset system shortcuts that can be assigned to macro pad buttons in simple mode.
enum SimpleAction: String, Codable, CaseIterable {
    case takeScreenshot
    case takeScreenshotArea
    case copy
    case paste
    case undo
    case redo
    case save
    case selectAll
    case newTab
    case closeTab
    case switchApp
    case spotlight
    case missionControl
    case showDesktop
    case lockScreen

    var displayName: String {
        switch self {
        case .takeScreenshot:     return "Screenshot"
        case .takeScreenshotArea: return "Screenshot Area"
        case .copy:               return "Copy"
        case .paste:              return "Paste"
        case .undo:               return "Undo"
        case .redo:               return "Redo"
        case .save:               return "Save"
        case .selectAll:          return "Select All"
        case .newTab:             return "New Tab"
        case .closeTab:           return "Close Tab"
        case .switchApp:          return "Switch App"
        case .spotlight:          return "Spotlight"
        case .missionControl:     return "Mission Control"
        case .showDesktop:        return "Show Desktop"
        case .lockScreen:         return "Lock Screen"
        }
    }

    var category: String {
        switch self {
        case .takeScreenshot, .takeScreenshotArea, .spotlight, .missionControl, .showDesktop, .lockScreen, .switchApp:
            return "System"
        case .copy, .paste, .undo, .redo, .save, .selectAll:
            return "Editing"
        case .newTab, .closeTab:
            return "Navigation"
        }
    }

    /// The virtual key code and modifier flags needed to simulate this shortcut.
    var keyCombo: (key: UInt16, flags: CGEventFlags) {
        switch self {
        case .takeScreenshot:
            // Cmd+Shift+3 — key 20 is '3'
            return (20, [.maskCommand, .maskShift])
        case .takeScreenshotArea:
            // Cmd+Shift+4 — key 21 is '4'
            return (21, [.maskCommand, .maskShift])
        case .copy:
            // Cmd+C — key 8
            return (8, .maskCommand)
        case .paste:
            // Cmd+V — key 9
            return (9, .maskCommand)
        case .undo:
            // Cmd+Z — key 6
            return (6, .maskCommand)
        case .redo:
            // Cmd+Shift+Z — key 6
            return (6, [.maskCommand, .maskShift])
        case .save:
            // Cmd+S — key 1
            return (1, .maskCommand)
        case .selectAll:
            // Cmd+A — key 0
            return (0, .maskCommand)
        case .newTab:
            // Cmd+T — key 17
            return (17, .maskCommand)
        case .closeTab:
            // Cmd+W — key 13
            return (13, .maskCommand)
        case .switchApp:
            // Cmd+Tab — key 48
            return (48, .maskCommand)
        case .spotlight:
            // Cmd+Space — key 49
            return (49, .maskCommand)
        case .missionControl:
            // Ctrl+Up — key 126
            return (126, .maskControl)
        case .showDesktop:
            // F11 — key 103
            return (103, CGEventFlags())
        case .lockScreen:
            // Ctrl+Cmd+Q — key 12
            return (12, [.maskControl, .maskCommand])
        }
    }

    /// All categories in display order.
    static var categories: [String] {
        ["System", "Editing", "Navigation"]
    }

    /// Actions filtered by category.
    static func actions(in category: String) -> [SimpleAction] {
        allCases.filter { $0.category == category }
    }
}
